### PR TITLE
Added option to specify socket type

### PR DESCRIPTION
**Pull request by bhaskarvk**
So that we can use SOCK_DGRAM instead of SOCK_RAW and avoid needing root privileges.

### CODE REVIEW

**Comment by bhaskarvk** [comment]
@octo Any interest in merging this PR. It allows to use the library w/o needing root priv.

**Comment by anarcat** [comment]
+1, this looks like a great idea... if i have a suggestion, it would be to keep the .gitignore change separate, otherwise this is fine...

**Comment by anarcat** [comment]
oh, and you should remove the 1.8.1 release tag from your fork - i think that those tags should be laid only by the source fork. :)

File: src/liboping.c [hunk 1]
```diff
@@ -1453,16 +1465,16 @@ int ping_host_add (pingobj_t *obj, const char *host)
...
 		}
 		else if (ai_ptr->ai_family == AF_INET6)
 		{
-			ai_ptr->ai_socktype = SOCK_RAW;
+			ai_ptr->ai_socktype = ai_hints.ai_socktype;
 			ai_ptr->ai_protocol = IPPROTO_ICMPV6;
 		}
 		else
-		{
+     		{
```
Comment (anarcat) [inline src/liboping.c]: avoid unrelated whitespace changes.

File: configure.ac [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 AC_PREREQ([2.65])
 AC_INIT([liboping],
-	[1.8.0],
+	[1.8.1],
```
Comment (anarcat) [inline configure.ac]: i would avoid changing the release number, otherwise to merge this MR, the author would absolutely need to make a release...

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,7 @@
+2015-04-03, Version 1.8.1:
```
Comment (anarcat) [inline ChangeLog]: i would avoid changing the release number, otherwise to merge this MR, the author would absolutely need to make a release...

File: src/liboping.c [hunk 4]
```diff
@@ -1338,6 +1340,16 @@ int ping_setopt (pingobj_t *obj, int option, void *value)
 		} /* case PING_OPT_MARK */
 		break;
 
+		case PING_OPT_SOCKTYPE:
+			obj->sockettype = *((int *) value);
```
Comment (anarcat) [inline src/liboping.c]: since this is basically a toggle, why pass a value to `-s`? and how is the user supposed to know the numeric value of `SOCK_RAW` or `SOCK_DGRAM`?

i would advise using `-d` to enable UDP instead.

Comment (anarcat) [inline src/liboping.c]: oh nevermind, i didn't see the parsing below...

still, i wonder why the argument is necessary.

**Comment by anarcat** [comment]
ah and hum. this doesn't actually work here:

```
[1029]anarcat@angela:liboping$ ./src/oping anarc.at
Adding host `anarc.at' failed: Operation not permitted
[1030]anarcat@angela:liboping1$ ./src/oping -s DGRAM anarc.at
Adding host `anarc.at' failed: Permission denied
[1031]anarcat@angela:liboping1$ ./src/oping -s RAW anarc.at
Adding host `anarc.at' failed: Operation not permitted
[1032]anarcat@angela:liboping1$ sudo ./src/oping anarc.at
PING anarc.at (72.0.72.144) 56 bytes of data.
56 bytes from anarc.at (72.0.72.144): icmp_seq=1 ttl=48 time=48.80 ms
^C
--- anarc.at ping statistics ---
1 packets transmitted, 1 received, 0.00% packet loss, time 48.8ms
RTT[ms]: min = 49, median = 49, p(95) = 49, max = 49
```

how do i actually use this?

File: src/liboping.c [hunk 5]
```diff
@@ -134,6 +134,7 @@ struct pingobj
 	double                   timeout;
 	int                      ttl;
 	int                      addrfamily;
+        int                      sockettype;
```
Comment (anarcat) [inline src/liboping.c]: whitespace is out of whack here.

File: src/liboping.c [hunk 6]
```diff
@@ -1453,16 +1465,16 @@ int ping_host_add (pingobj_t *obj, const char *host)
 
 		if (ai_ptr->ai_family == AF_INET)
 		{
-			ai_ptr->ai_socktype = SOCK_RAW;
+			ai_ptr->ai_socktype = ai_hints.ai_socktype;
 			ai_ptr->ai_protocol = IPPROTO_ICMP;
```
Comment (anarcat) [inline src/liboping.c]: actually, the more i look at this, the more i don't think this can work: just changing `ai_socktype` without changing `ai_protocol` will not work. I believe this is why i get errors when testing the code: you can't just send ICMP packets on a UDP socket, it makes no sense at all.

and even if you managed somehow to open such a socket somewhere, you would still be creating ICMP datagrams elsewhere in the code and trying to send that over that UDP socket. really, this is a much larger change that needs to happen here...

**Comment by bhaskarvk** [comment]
I'll close this for now. I haven't worked on this in a while, I'll double check but I did get it to work on my Mac last time I tried.
I'll fix all the issues you mentioned and open a new PR when ready.

**Comment by bhaskarvk** [comment]
I'm reopening coz I figured out what needs to be done to make it work correctly. I'll fix the code indent issues and convert the command-line option to a flag.

SOCK_DGRAM works for *BSD / OSX Kernels, and Linux Kernels (with a special sysctl setting).

## For OS-X/*BSD 

From https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man4/icmp.4.html

```
 Non-privileged ICMP
     ICMP sockets can be opened with the SOCK_DGRAM socket type without requiring root privileges. The syn-opsis synopsis
     opsis is the following:

     socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP)

     Datagram oriented ICMP sockets offer a subset of the functionality available to raw ICMP sockets. Only
     IMCP request messages of the following types can be sent: ICMP_ECHO, ICMP_TSTAMP or ICMP_MASKREQ.  The
     code field must be the value zero (0).  The minimal length of an ICMP message request is eight (8)
     octets.
```

So under OS-X and BSD kernels it should work as is.

More discussion at 
https://stackoverflow.com/questions/1189389/python-non-privileged-icmp

## For Linux you need to do make some sysctl changes 
```
# Ubuntu Touch r14
net.ipv4.ping_group_range ="0   2147483647"

# Ubuntu 15.04 Vivid Vervet
net.ipv4.ping_group_range ="1   0"

# Fedora 21
net.ipv4.ping_group_range ="1   0"
```
Some notes on this
https://sturmflut.github.io/linux/ubuntu/2015/01/17/unprivileged-icmp-sockets-on-linux/
https://stackoverflow.com/questions/8290046/icmp-sockets-linux/20105379#20105379
And the corresponding patch that went in to the Linux Kernel
https://lkml.org/lkml/2011/5/10/389

**Comment by anarcat** [comment]
i see - so the following would need to be added to sysctl.conf for this to work in linux:

```
net.ipv4.ping_group_range = 0 4294967295
```

is that correct? If so i think this should be added to the manpage at least.

**Comment by bhaskarvk** [comment]
Yes, depending upon the Linux distro either 
`net.ipv4.ping_group_range ="0   2147483647"` OR `net.ipv4.ping_group_range ="0   1"`.
I'll make sure to add it to the man page.

**Comment by anarcat** [comment]
alright, i can confirm that `sysctl net.ipv4.ping_group_range="0 2147483647"` works here in Debian Jessie (Linux 3.16), insofar as it can create the socket and send packets. However, it gets a timeout instead of success:

```
[1086]anarcat@angela:liboping$ ./src/oping -s DGRAM -c 1 koumbit.org
PING koumbit.org (199.58.80.86) 56 bytes of data.
echo reply from koumbit.org (199.58.80.86): icmp_seq=1 timeout

--- koumbit.org ping statistics ---
1 packets transmitted, 0 received, 100.00% packet loss, time 0.0ms
[1087]anarcat@angela:liboping$ sudo ./src/oping -c 1 koumbit.org
PING koumbit.org (199.58.80.86) 56 bytes of data.
56 bytes from koumbit.org (199.58.80.86): icmp_seq=1 ttl=55 time=23.88 ms

--- koumbit.org ping statistics ---
1 packets transmitted, 1 received, 0.00% packet loss, time 23.9ms
RTT[ms]: min = 24, median = 24, p(95) = 24, max = 24
```

so things are not quite working in Linux just yet.

**Comment by bhaskarvk** [comment]
I'll see what I can do for the Linux part.

**Comment by bhaskarvk** [comment]
I got timeout on Linux too (same Debian Jessie) , But this https://stackoverflow.com/a/20105379/507484 works. So it's only a matter of figuring out what needs to change in order to make it work under Linux.
I'll see what I can do.

**Comment by anarcat** [comment]
thanks!

**Comment by bhaskarvk** [comment]
This is the strace of the program in  https://stackoverflow.com/a/20105379/507484, which works

```
socket(PF_INET, SOCK_DGRAM, IPPROTO_ICMP) = 3
sendto(3, "\10\0\0\0\322\4\0\0hello", 13, 0, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("63.88.73.24")}, 16) = 13
fstat(1, {st_mode=S_IFREG|0644, st_size=1924, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd5587c5000
select(4, [3], NULL, NULL, {3, 0})      = 1 (in [3], left {2, 989959})
recvfrom(3, "\0\0\274\34\0\21\0\0hello", 2048, 0, NULL, 0x7fffb52ac24c) = 13
write(1, "Sent ICMP\n\nICMP Reply, id=0x4d2,"..., 49Sent ICMP
```

While this is the strace of oping which is timing out
```
socket(PF_INET, SOCK_DGRAM, IPPROTO_ICMP) = 3
setsockopt(3, SOL_SOCKET, SO_TIMESTAMP, [1], 4) = 0
setsockopt(3, SOL_IP, IP_RECVTOS, [1], 4) = 0
setsockopt(3, SOL_IP, IP_RECVTTL, [1], 4) = 0
setsockopt(3, SOL_IP, IP_TTL, [64], 4)  = 0
setsockopt(3, SOL_IP, IP_TOS, "\0", 1)  = 0
getuid()                                = 900
setuid(900)                             = 0
rt_sigaction(SIGINT, {0x4029d0, [], SA_RESTORER, 0x7f616f42b180}, NULL, 8) = 0
fstat(1, {st_mode=S_IFREG|0644, st_size=33234, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f616fec1000
sendto(3, "\10\0\262\337e\216\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0libo"..., 84, 0, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("199.58.80.86")}, 16) = 84
select(4, [3], NULL, [3], {1, 0})       = 1 (in [3], left {0, 973846})
recvmsg(3, {msg_name(0)=NULL, msg_iov(1)=[{"\0\0 \\\0\22\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0libo"..., 4096}], msg_controllen=80, {cmsg_len=32, cmsg_level=SOL_SOCKET, cmsg_type=0x1d /* SCM_??? */, ...}, msg_flags=0}, 0) = 84
select(4, [3], NULL, [3], {0, 973568})  = 0 (Timeout)
write(1, "PING koumbit.org (199.58.80.86) "..., 113PING koumbit.org (199.58.80.86) 56 bytes of data.
echo reply from koumbit.org (199.58.80.86): icmp_seq=1 timeout
) = 113
```

As far as I can tell the only real difference is the use of recvmsg in oping and recvfrom in the working code. Otherwise they look almost same barring some setsockopt calls in oping.

The oping code works perfectly under Mac, but just not Linux. Strange indeed.

**Comment by anarcat** [comment]
the sendmsg packets are different as well... the binary data sent on the wire is different. not sure what that means. also, there is an easier way of making oping non-suid: use capabilities. i filed a bug in debian about this: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=804784

**Comment by bhaskarvk** [comment]
Yes the packets are different because I was just testing with the default payload of the two programs, which happen to be different. I don't think I've enough bandwidth this week to figure this out, but I'll try to make this work sometime next week. Thanks for your pointers.

As to your Bug upstream, my primary interest is using liboping from R/python, capabilities won't be of much help there. Using SOCK_DGRAM allows me to do just that w/o having to set SUID or capabilities flag on the R / python binary.

**Comment by octo** [comment]
Hi @bhaskarvk and @anarcat,

thank you very much for this patch and thanks @anarcat for the review comments, it's much appreciated!

IIuc this change uses a BSD feature to use unprivileged DGRAM sockets to send ICMP echos. Something similar can be achieved on Linux by setting some sysctls. I think that would be an awesome feature to use for *noping*.

Ideally, I the default would be an automatic selection, i.e. use DGRAM as the default when running on a BSD kernel (this could be a preprocessor `#if` for now) and check the sysctl settings when on Linux.

Best regards,
—octo

**Comment by bhaskarvk** [comment]
I really don't have b/w to work on this for now, so I'm closing this PR.
If I figure out how to make this work consistently across BSD/Linux/MacOS then I'll re open.
